external/js/cothority/lib/skipchain/skipchain.js

```diff
@@ -0,0 +1,217 @@
+      // XXX  somewhat hackyish but sets a realistic upper bound
+      const initLength = client.lastRoster.length;
+      var nbErr = 0;
+      while (nbErr < initLength) {
+        // fetches the data with the current roster
+        client.socket = new net.RosterSocket(client.lastRoster, skipchainID);
+        var data = null;
+        try {
+          data = yield client.socket.send(requestStr, responseStr, request);
+        } catch (err) {
+          return Promise.reject(err);
+        }
+        // verifies it's all correct
+        var lastBlock, err;
+        [lastBlock, err] = client.verifyUpdateChainReply(data);
+        if (!err) {
+          // tries again with random conodes
+          nbErr++;
+        }
+        // update the roster
+        client.lastRoster = identity.Roster.fromProtobuf(lastBlock.roster);
+        client.lastID = lastBlock.hash;
+        // if there is nothing new stop !
+        if (!lastBlock.forward || lastBlock.forward.length == 0) {
+          // no forward block means it's the latest block
+          return Promise.resolve(lastBlock);
+        }
+      }
+      return Promise.reject(nbErr + " occured retrieving the latest block...");
+    });
+    return fn(this);
+  }
+
+  /**
+   * verifies if the list of skipblock given is correct and if it links with the last know given skipblockID.
+   *
+   * @param {Uint8Array} lastID last know skipblock ID
+   * @param {GetUpdateChainReply} updateChainReply the response from a conode containing the blocks
+   * @returns {(SkipBlock,err)} the most recent valid block in the chain, or an error
+   */
+  verifyUpdateChainReply(updateChainReply) {
+    console.log("Verifying update...");
+    const blocks = updateChainReply.update;
+    if (blocks.length == 0) {
+      return [null, new Error("no block returned in the chain")];
+    }
+    // first verify the first block is the one we know
+    const first = blocks[0];
+    const id = new Uint8Array(first.hash);
+    if (!misc.uint8ArrayCompare(id, this.lastID)) {
+      return [null, new Error("the first ID is not the one we have")];
+    }
+
+    if (blocks.length == 1) {
+      return [first, null];
+    }
+    // then check the block links consecutively
+    var currBlock = first;
+    for (var i = 1; i < blocks.length; i++) {
+      const nextBlock = blocks[i];
+
+      const forwardLinks = currBlock.forward;
+      if (forwardLinks.length == 0)
+        return [null, new Error("No forward links included in the skipblocks")];
+
+      // only take the highest link since we move "as fast as possible" on
+      // the skipchain, i.e. we skip the biggest number of blocks
+      const lastLink = forwardLinks[forwardLinks.length - 1];
+      // XXX to change later to source_hash, dest_hash, dst_roster_id
+      const message = nextBlock.hash;
+      const roster = identity.Roster.fromProtobuf(currBlock.roster);
+      //var err = this.verifyForwardLink(roster, message, lastLink);
```

Comment:
Is this left commented out on purpose?

Reply:
Yes because I'd like to advance on the Agora stuff as quickly as possible so the code that verifies is there but untested. It will be once I have more time.